APE_Black_Formatting.rst

```diff
@@ -0,0 +1,308 @@
+
+date-last-revised: 2022 August 25
+
+date-accepted: <replace with accepted date>
+
+type: Standard Track
+
+status: Discussion
+
+
+Abstract
+--------
+
+This APE proposes that the `astropy core package
+<https://github.com/astropy/astropy>`_ adopt Black_ for formatting code.
+
+Detailed description
+--------------------
+
+Foreword
+^^^^^^^^^
+
+This APE presents a rationale that code formatting with Black_ can be beneficial
+both for the ``astropy`` code development process and for the community.
+Hereafter we refer to the astropy core package as ``astropy``, noting that it is
+distinct from the Astropy project as a whole and coordinated and affiliated
+Astropy packages. While coordinated or affiliated packages may choose to adopt
+this APE's conventions in the future, this APE is concerned solely with the core
+package itself.
+
+Black_ is a fast and reliable Python automated code formatter supported by the
+Python Software Foundation. The project motto "The uncompromising code
+formatter" reflects that the code is strongly opinionated about formatting and
+there are very few configuration options. From their documentation, "By using
+*Black*, you agree to cede control over minutiae of hand-formatting. In return,
+Black gives you speed, determinism, and freedom from ``pycodestyle`` nagging
+about formatting."
+
+Code formatting is controversial and the `astropy-dev thread`_ about Black
+generated a spirited discussion but did not result in a clear consensus. This
+strong division of opinions about automated code formatting is common and has
+been documented in other projects, for instance in `Django DEP-0008`_.
+
+Other projects, including NumPy and SciPy, have considered using Black for
+formatting (see e.g. `SciPy PR #14330
+<https://github.com/scipy/scipy/pull/14330>`_). Neither of those projects felt
+that auto-formatting with the current version of Black (circa 2022) was a good
+choice for their codebase, with particular emphasis on needing improvements
+in formatting of mathematical expressions.
+
+We do not expect unanimous agreement on this APE but do hope to demonstrate the
+value of Black for the collective needs of the community and Astropy Project.
+
+.. _astropy-dev thread: https://groups.google.com/g/astropy-dev/c/6cRJCMgaFyM/
+
+Note that ideas and some text for this APE have been adapted from a similar
+enhancement proposal for Django `Django DEP-0008`_.
+
+Rationale
+^^^^^^^^^
+
+**Consistent and readable code formatting** reduces effort in both code review
+and in future maintenance. Code is read far more often than it is written, so it
+makes sense to invest effort to ensure that the code is easy to read later.
+
+Ensuring that code formatting meets ``astropy`` standards is currently done by
+applying CI checks based on :pep:`8`. Learning and consistently applying all
+these rules is not easy. There are at least 200 commits in ``astropy`` which
+were solely address PEP8 violations, so this is a real issue.
```

Comment:
This claim about more than 200 commits fixing code style originates from the mailing list, and the number was obtained as follows (using the current `main` e4993fffb54e19b04bc4e9af084984650bc0a46f):
```shell
$ git log --grep="PEP\s*8" --pretty=oneline | wc -l
229
```
A more thorough search reveals an even bigger problem:
```shell
$ git log --grep="PEP\s*8" --grep="pep\s*8" --grep="flake\s*8" --grep="code\s*style" --pretty=oneline | wc -l
409
```

Reply:
But these are due to the non-enforcing nature of formatting and not due to not using black and previous resistance of being more strict with pep8 enforcement. A properly set up editor with flake8/pre-commit runs with it, etc would have just as well avoided the extra commits.

Reply:
Are you talking about autopep8 in pre commit? Or something else?

Reply:
yes, for example, that. Or a properly set up editor that picks up on the pycodestyle/flake8/etc config from the repo.

Or simply maintainers pressure to make contributors squash commits down to logical chunks.

Reply:
@bsipocz - about `autopep8` as an autoformatter -- that loses the advantage of the style uniformity and readability improvements that black brings (in the opinion of the proponents of this APE). Basically for something like below (and *many* other cases in the astropy code base), autopep8 is perfectly happy with the original because it is indeed PEP8-compliant.

<img src="https://user-images.githubusercontent.com/348089/188288311-984cb1e9-0a49-45a4-b4d8-6bcc2632ad53.png" width=500>

Reply:
The APE use the argument that we need black as we have 200 commits of pep8 fixes. And all in saying, if we could have been strict with enforcing formatting before we would have had less.

Now, this particular example you bring: The original 3 lines are perfectly readable for me, especially in the context it's used, much more than the proposed 11. And I've brought examples with maths and nested containers where it's difficult to argue that the black version is really acceptable.

Reply:
@bsipocz - fair enough and thanks for all your input. This APE starts with a Foreword that mentions not expecting consensus. I think we have realized that expectation and I do not see any way to further modify this APE in a way that will be acceptable to both you and the APE authors.

This PR has been open for discussion for 12 days now, with not much activity in the last week. Other comments have neutral to positive, mostly focusing on improving the APE (wording, rationale, and implementation). I'd like to ask the (unconflicted) CoCo members @hamogu @kelle @adrn @mwcraig  for direction on going forward from here.

Reply:
Thanks for the ping @taldcroft -- as @hamogu noted in his email, this is on our agenda to discuss on Monday.